Respect repomapping on distros other than RHEL

matejmatuska opened the pull request:
This series of commits together enable the use of repomapping on distros other than RHEL. Without these changes the `--enablerepo` option had to be used to specify target repos and they were treated as custom repos.

Extended `distro` shared library by new functions:
* `get_distro_repoids`
* `get_target_distro_repoids`

Some other more or less related changes:
- cleanup unused code path in setuptarget repos
- do not run checktargetrepos on distros other than RHEL
- I added tests for the existing `get_distribution_data()` function in distro library


### Deprecation
* Deprecates the `RHELTargetRepository` model - replaced by `DistroTargetRepository`

pirat89 commented on repos/system_upgrade/common/models/targetrepositories.py:
```diff
@@ -32,6 +41,8 @@ class TargetRepositories(Model):
     are missing, upgrade can still continue.
```
this docstring needs to be updated as well.

matejmatuska commented on repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:
```diff
@@ -962,9 +981,11 @@ def extract_repoid_from_line(line):
 
     api.current_logger().debug(
-        'The following repofiles are considered as provided by RedHat: {0}'.format(' '.join(rh_repoids))
+        "The following repofiles are considered as provided by the distribution: {0}".format(
+            " ".join(distro_repoids)
```
Done

matejmatuska replied:
I moved the logging from the function to the caller site. Since the function is now in the distro library I think the caller should log the results if they want.

matejmatuska commented:
I will mark this as ready, but have yet to test it thoroughly.

matejmatuska commented:
Okay so I found some problems during e2e testing

1) This has also been caught by testing farm tests. On RHEL with --no-rhsm, the upgrade was incorrectly inhibited due to missing base repos.
I fixed this and added a unit test covering this in a new commit for now.

2) The `TargetRepositories.rhel_repos` was being filled on distros other than RHEL, fixed in another commit.

3) Another problem I found is that there is a usage of deprecated model report produced:
```
Risk Factor: medium
Title: Usage of deprecated Model "RHELTargetRepository" at /usr/lib/python3.6/site-packages/leapp/models/fields/__init__.py:565
Summary: This model is deprecated, use DistroTargetRepository instead.
Since: 2025-07-23
Location: /usr/lib/python3.6/site-packages/leapp/models/fields/__init__.py:565
Near:         return self._model_type(init_method='to_model', **value)
```
however if I add `suppress_deprecation()` to the class it causes and error:
```
INTERNALERROR>   File "/repocopy/tut/lib/python3.9/site-packages/leapp/utils/deprecation.py", line 28, in decorator
INTERNALERROR>     raise ValueError(
INTERNALERROR> ValueError: The suppress_deprecation decorator has been used on a class which does not contain the process method. The decorator can be used only for classes with that method (e.g. classes derived from Actor). Use the decorator on the affected methods instead of the current class.
``` 
@pirat89 do we have some other way to suppress deprecation on a Model?

matejmatuska commented:
> > It looks like the trigger is that in consume() the Model objects are recreated from JSON.
> 
> yes, it's typical. There is not problem with that.
> 
> > I think there is no way to suppress this without changes in the framework. I tried putting the suppression on _any_custom_repo_defined() but it is probably too "high" in the callstack so it doesn't take effect.
> 
> the stack is not important. well - it's important just in case of messages (that's why the stack option has been implemented as a workaround, which in some cases does not have to be ideal - like in this one, I agree; we could update this in future, but as it is not so often issues, it's probably not worthy). Trying the suppression, it works for me. Do not forget that you want to suppress the warning for `RHELTargetRepository` , not `TargetRepository` (so you need to also import the model).

Yeah I did suppress it that way and imported as well, let me try again, maybe I did something wrong.

matejmatuska commented:
## Tests

### CS test
<details><summary>Details</summary>
<p>
Tested on CS on 9->10 upgrade, `rt` repo enabled.

#### Without this patch, using --enablerepo:
```
[root@localhost ~]# leapp upgrade --debug --enablerepo baseos --enablerepo appstream
[root@localhost ~]# leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-04T14:36:37.319428Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [
        {
            "baseurl": null,
            "enabled": true,
            "name": null,
            "repoid": "appstream"
        },
        {
            "baseurl": null,
            "enabled": true,
            "name": null,
            "repoid": "baseos"
        }
    ],
    "rhel_repos": [
        {
            "repoid": "appstream"
        },
        {
            "repoid": "baseos"
        },
        {
            "repoid": "rt"
        }
    ]
}
######################################################################
```
#### With this patch:
The repoids are correctly contained only in `TargetRepositories.distro_repos`, no need for `--enablerepo`:
```
[root@localhost ~]#  leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-15T10:23:21.037344Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [],
    "distro_repos": [
        {
            "repoid": "appstream"
        },
        {
            "repoid": "baseos"
        },
        {
            "repoid": "rt"
        }
    ],
    "rhel_repos": []
}
######################################################################
```
The distro provided repoids are properly obtained:
```
2025-09-15 12:23:24.71  INFO     PID: 144863 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: The following repoids are considered as provided by the 'centos' distribution:
    - appstream
    - appstream-debuginfo
    - appstream-source
    - baseos
    - baseos-debuginfo
    - baseos-source
    - crb
    - crb-debuginfo
    - crb-source
    - extras-common
    - extras-common-source
    - highavailability
    - highavailability-debuginfo
    - highavailability-source
    - nfv
    - nfv-debuginfo
    - nfv-source
    - resilientstorage
    - resilientstorage-debuginfo
    - resilientstorage-source
    - rt
    - rt-debuginfo
    - rt-source
```
And the gathered repoids are also correct:
```
[root@localhost ~]# grep Gathered /var/log/leapp/leapp-preupgrade.log
2025-09-15 12:23:24.146 DEBUG    PID: 144863 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: Gathered target repositories: appstream, baseos, rt
```
</p>
</details> 


### Test that distro_repoids is used on RHEL and there is no regression
#### 9->10

<details><summary>Details</summary>
<p>

##### RHSM
RHEL 9.6->10.0 with RHSM (there are also the custom leapp testing repos):
```
[root@localhost ~]# leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-04T13:17:27.013238Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [
        {
            "baseurl": "censored",
            "enabled": false,
            "name": "APPSTREAM",
            "repoid": "APPSTREAM"
        },
        {
            "baseurl": "censored",
            "enabled": false,
            "name": "BASEOS",
            "repoid": "BASEOS"
        }
    ],
    "distro_repos": [
        {
            "repoid": "rhel-10-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-10-for-x86_64-baseos-rpms"
        }
    ],
    "rhel_repos": [
        {
            "repoid": "rhel-10-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-10-for-x86_64-baseos-rpms"
        }
    ]
}
######################################################################
```
```
There is a lot of provided repos, showing only appstream and baseos here:
[root@leapp-20250915124455 ~]# grep -A 100  'provided\ by'  /var/log/leapp/leapp-preupgrade.log  | grep '\s\s\s-\s'.*appstream-rpms
    - rhel-10-for-x86_64-appstream-rpms
[root@leapp-20250915124455 ~]# grep -A 100  'provided\ by'  /var/log/leapp/leapp-preupgrade.log  | grep '\s\s\s-\s'.*baseos-rpms
    - rhel-10-for-x86_64-baseos-rpms
```
```
[root@leapp-20250915124455 ~]# grep Gathered /var/log/leapp/leapp-preupgrade.log
2025-09-15 11:13:59.185 DEBUG    PID: 57328 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: Gathered target repositories: rhel-10-for-x86_64-appstream-rpms, rhel-10-for-x86_64-baseos-rpms, APPSTREAM, BASEOS
```

##### no RHSM
```
[root@leapp-20250915124455 ~]# leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-15T10:50:48.148756Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [
        {
            "baseurl": "<censored>",
            "enabled": false,
            "name": "APPSTREAM",
            "repoid": "APPSTREAM"
        },
        {
            "baseurl": "<censored>",
            "enabled": false,
            "name": "BASEOS",
            "repoid": "BASEOS"
        }
    ],
    "distro_repos": [
        {
            "repoid": "rhel-10-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-10-for-x86_64-baseos-rpms"
        }
    ],
    "rhel_repos": [
        {
            "repoid": "rhel-10-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-10-for-x86_64-baseos-rpms"
        }
    ]
}
######################################################################
```
```
[root@leapp-20250915124455 ~]# grep 'provided\ by'  /var/log/leapp/leapp-preupgrade.log
2025-09-15 10:50:50.372 INFO     PID: 43607 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: The following repoids are considered as provided by the 'rhel' distribution: []
```
```
[root@leapp-20250915124455 ~]# grep Gathered /var/log/leapp/leapp-preupgrade.log
2025-09-15 10:50:50.451 DEBUG    PID: 43607 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: Gathered target repositories: APPSTREAM, BASEOS
```

</p>
</details> 

#### 8->9
<details><summary>Details</summary>
<p>

##### RHSM
RHEL 8.10 to 9.6 with RHSM and the custom testing repos:
```
[root@leapp-20250904144519 ~]# leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-04T13:49:04.158882Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [
        {
            "baseurl": "censored",
            "enabled": false,
            "name": "APPSTREAM",
            "repoid": "APPSTREAM"
        },
        {
            "baseurl": "censored",
            "enabled": false,
            "name": "BASEOS",
            "repoid": "BASEOS"
        }
    ],
    "distro_repos": [
        {
            "repoid": "rhel-9-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-9-for-x86_64-baseos-rpms"
        }
    ],
    "rhel_repos": [
        {
            "repoid": "rhel-9-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-9-for-x86_64-baseos-rpms"
        }
    ]
}
######################################################################
```
There is a lot more, showing just these:
```
[root@leapp-20250905170450 ~]# grep -A 300  'provided\ by'  /var/log/leapp/leapp-preupgrade.log  | grep '\s\s\s-\s'.*baseos-rpms
    - rhel-9-for-x86_64-baseos-rpms
[root@leapp-20250905170450 ~]# grep -A 300  'provided\ by'  /var/log/leapp/leapp-preupgrade.log  | grep '\s\s\s-\s'.*appstream-rpms
    - rhel-9-for-x86_64-appstream-rpms
```
```
2025-09-15 11:06:36.935 DEBUG    PID: 205607 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: Gathered target repositories: rhel-9-for-x86_64-appstream-rpms, rhel-9-for-x86_64-baseos-rpms, APPSTREAM, BASEOS
```

##### no RHSM

RHEL 8.10->9.6, no RHSM:
The custom repos are where they should be, and distro_repos (and rhel_repos) are correctly obtained from repommap data.
```
[root@leapp-20250905170450 ~]# leapp-inspector messages --type TargetRepositories
######################################################################
                          PRODUCED MESSAGES
######################################################################
Stamp: 2025-09-15T10:22:55.568992Z
Actor: setuptargetrepos
Phase: FactsCollection
Type: TargetRepositories
Message_data:
{
    "custom_repos": [
        {
            "baseurl": "<censored>",
            "enabled": false,
            "name": "APPSTREAM",
            "repoid": "APPSTREAM"
        },
        {
            "baseurl": "<censored>",
            "enabled": false,
            "name": "BASEOS",
            "repoid": "BASEOS"
        }
    ],
    "distro_repos": [
        {
            "repoid": "rhel-9-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-9-for-x86_64-baseos-rpms"
        }
    ],
    "rhel_repos": [
        {
            "repoid": "rhel-9-for-x86_64-appstream-rpms"
        },
        {
            "repoid": "rhel-9-for-x86_64-baseos-rpms"
        }
    ]
}
######################################################################
```
```
[root@leapp-20250905170450 ~]# grep 'provided\ by'  /var/log/leapp/leapp-preupgrade.log
2025-09-15 10:45:48.267 INFO     PID: 187993 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: The following repoids are considered as provided by the 'rhel' distribution: []
```
```
[root@leapp-20250905170450 ~]# grep Gathered /var/log/leapp/leapp-preupgrade.log
2025-09-15 10:23:01.100 DEBUG    PID: 178860 leapp.workflow.TargetTransactionFactsCollection.target_userspace_creator: Gathered target repositories: APPSTREAM, BASEOS
```

</p>
</details>

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +18,56 @@ def get_distribution_data(distribution):
             details={'Problem': 'Distribution {} was not found in {}.'.format(distribution, distributions_path)})
+
+
+def get_distro_repofiles(distro=None):
```
make sense. in such a case remove also the line 33. but let me think about it yet a little bit. also, if we do not expect that someone will want to use it, we can make the function private.

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +18,56 @@ def get_distribution_data(distribution):
+    return distro_provided_repofiles[distro]
+
+
+def get_distro_repoids(context, distro=None):
```
if the functionality of `get_distro_repofiles` is modified, the same should happen here as well.

matejmatuska replied:
Yes I mentioned that in the original comment.

pirat89 replied:
@matejmatuska actually, with this setup, if you run it on AlmaLinux it would crash right away. So we should put here already almalinux repofiles also. even in case it would not be complete set, it will at least not crash.

```
/etc/yum.repos.d/almalinux-appstream.repo
/etc/yum.repos.d/almalinux-baseos.repo
/etc/yum.repos.d/almalinux-crb.repo
/etc/yum.repos.d/almalinux-extras.repo
/etc/yum.repos.d/almalinux-highavailability.repo
/etc/yum.repos.d/almalinux-nfv.repo
/etc/yum.repos.d/almalinux-plus.repo
/etc/yum.repos.d/almalinux-resilientstorage.repo
/etc/yum.repos.d/almalinux-rt.repo
/etc/yum.repos.d/almalinux-sap.repo
/etc/yum.repos.d/almalinux-saphana.repo
```


@yuravk hi, could you check that these are all files you would like to list here? tl;dr; this PR is prerequisite to actually use correctly repomap.json file on centos-like distributions.

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +18,56 @@ def get_distribution_data(distribution):
+
+    repofiles = repofileutils.get_parsed_repofiles(context)
+    distro_repoids = []
+    for rfile in repofiles:
+        if rfile.file in get_distro_repofiles(distro) and rfile.data:
+            distro_repoids.extend([repo.repoid for repo in rfile.data])
```
i am thinking that we shuld actually check whether such repofiles are present on in the given context. what if the installed container does not contain installed a package with required repofiles? it would be nice to have logged whether the file (or some of files) are actually missing - for the debugging purposes.

matejmatuska replied:
I think we synced about this outside of GitHub without any real conclusion, but I think it would be beneficial to have at least the logs.

pirat89 reviewed (changes requested):
I have not finished the review yet, but sharing current findings.

matejmatuska commented on repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:
```diff
@@ -986,15 +979,29 @@ def gather_target_repositories(context, indata):
+            )
+        )
+    else:
+        api.current_logger().info(
+            "The following repoids are considered as provided by the '{}' distribution: []".format(
+                get_distro_id()
+            )
+        )
```
What's incorrect? It's intentionally like this because if there was only the message in the `if` then when `repoids` are empty the message would contain a newline with the separator.

matejmatuska replied:
I can make it a little nicer and less duplicated though.

matejmatuska replied:
I like it to be the same regardless of whether there are repoids because it's more grepable.

pirat89 replied:
TBH I do not like it, but if you want it this way, remove the if statement so it's more clear that this is really intended. like:

```
FMT_LIST_SEPARATOR = '\n    - '                                                 
                                                                                
                                                                                
def _formatted_list_output(input_list, sep=FMT_LIST_SEPARATOR):                 
    return ['{}{}'.format(sep, item) for item in input_list]                    
                                                              

....
    api.current_logger().info(
        "The following repoids are considered as provided by the '{}' distribution:{}"
        .format(
            get_distro_id(), _formatted_list_output(distro_repoids)
        )
    )
```

But I would honestly found more helpful to see a warning log with `No repoids provided by the {} distribution has been discovered` when the list is empty.

matejmatuska commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +18,56 @@ def get_distribution_data(distribution):
+            "/etc/yum.repos.d/centos-addons.repo",
+        ],
+    }
+    return distro_provided_repofiles[distro]
```
Yeah I would make it private.

pirat89 replied:
@matejmatuska actually, with this setup, if you run it on AlmaLinux it would crash right away. So we should put here already almalinux repofiles also. even in case it would not be complete set, it will at least not crash.

/etc/yum.repos.d/almalinux-appstream.repo
/etc/yum.repos.d/almalinux-baseos.repo
/etc/yum.repos.d/almalinux-crb.repo
/etc/yum.repos.d/almalinux-extras.repo
/etc/yum.repos.d/almalinux-highavailability.repo
/etc/yum.repos.d/almalinux-nfv.repo
/etc/yum.repos.d/almalinux-plus.repo
/etc/yum.repos.d/almalinux-resilientstorage.repo
/etc/yum.repos.d/almalinux-rt.repo
/etc/yum.repos.d/almalinux-sap.repo
/etc/yum.repos.d/almalinux-saphana.repo

@yuravk hi, could you check that these are all files you would like to list here? tl;dr; this PR is prerequisite to actually use correctly repomap.json file on centos-like distributions.

pirat89 replied:
@matejmatuska btw, it seems the list of repofiles is not constistant across major versions. so we will need the list version.

yuravk replied:
@pirat89 , @matejmatuska yes the repositories list is correct, but with the following executions:
- There is no _ResilientStorage_ repository in **AlmaLinux 10.x**
- _RT_ and _NFV_ repositories are available only for `x86_64` both in **AlmaLinux 9.x** and **10.x**

I believe both cases are true for the **CentOS Stream** as well.

Thank you

matejmatuska commented on repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:
```diff
@@ -885,13 +869,24 @@ def get_copy_location_from_copy_in_task(context_basepath, copy_task):
+    :return: A list of repoids provided by distribution
+    :rtype: List[String]
     """
+    if get_distro_id() == 'rhel' and rhsm.skip_rhsm():
+        return []
```
True and also true about the docstring, I will fix this.

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +18,56 @@ def get_distribution_data(distribution):
+    """
+    distro = distro or get_distro_id()
+    distro_provided_repofiles = {
+        "rhel": [
```
mixture of quotes and apostrophes in the file. use please just one of them. preferably apostrophes if these are used already originally in the file.

matejmatuska commented on repos/system_upgrade/common/actors/targetuserspacecreator/libraries/userspacegen.py:
```diff
@@ -885,13 +869,24 @@ def get_copy_location_from_copy_in_task(context_basepath, copy_task):
+
+    On RHEL: RH repositories are provided either by RHSM or are stored in the expected repo file provided by
+             RHUI special packages (every cloud provider has itw own rpm).
+             If RHSM is skipped there are no repoids.
```
Removed.

karolinku commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +19,199 @@ def get_distribution_data(distribution):
+    )
+
+    for repofile, archs in enumerate(version_repofiles):
+        if archs is str and archs is all:
```
I tested the code on C9 VM. I seems like this line will not work, failing with ValueError as those conditions are not met. `archs is str` will always be False, because `is` checks for object identity, not type. Also `archs is all` is not checking string content, should be `if archs == 'all'`.

matejmatuska commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +19,206 @@ def get_distribution_data(distribution):
+    'centos': {
+        '8': {
+            # TODO is this true on all archs?
+            'CentOS-Linux-AppStream.repo': 'all',
```
The `'all'` could be changed for empty list or perhaps `['noarch']` or something else.

matejmatuska commented:
/packit retest-failed

pirat89 commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +19,206 @@ def get_distribution_data(distribution):
+            '/etc/yum.repos.d/almalinux-highavailability.repo': 'all',
+            '/etc/yum.repos.d/almalinux-plus.repo': 'all',
+            '/etc/yum.repos.d/almalinux-sap.repo': 'all',
+            '/etc/yum.repos.d/almalinux-saphana.repo': 'all',
+            # RT and NFV are only on x86_64 on almalinux 10
+            '/etc/yum.repos.d/almalinux-nfv.repo': ['x86_64'],
```
for architectures, you please rather arch "constants" defined in the  `leapp.libraries.common.config.architecture` library. regarding all, what about using `ARCH_SUPPORTED`  instead? you can then simplify the logic. Also I would suggest to have one type of value, instead of str / list.

(note that `ARCH_SUPPORTED` is tuple)

yuravk commented on repos/system_upgrade/common/libraries/distro.py:
```diff
@@ -16,3 +20,191 @@ def get_distribution_data(distribution):
+
+    distro_repoids = []
+    for rfile in repofiles:
+        if rfile.file in distro_repofiles:
```
I'm very sorry for the late comment.

With the current implementation AlmaLinux 8 -> 9 is inhibited on `target_userspace_creator` with `_inhibit_if_no_base_repos`:
```
Cannot find required basic RHEL target repositories.
```
The reason is the 8-th `/etc/yum.repos.d/almalinux.repo` doesn't match here to any of pre-defined 9-th `repofiles`.
